### PR TITLE
Initial source code

### PR DESCRIPTION
**Pull request by LukeCarrier**
Tired (will open issues after merging):
1. XML support for the WS client, since we don't have a `serialise()` implementation for it. Given that the client is currently unused, I suggest we avoid this for now. (#2)
2. The issue with XML request payloads containing tags with duplicate names inside of `external_multiple_structure` values still persists. The current mitigation is to just add a unique identifier to the end of the tag, e.g. an int we just increment each time. This is left to the developer. (#3)

Wired:
1. Parsing loaded requests with bodies, given the appropriate HTTP method (`PUT`!).
2. Basic WS client.

Inspired: cleaned up Git history.

### CODE REVIEW

File: tests/server_test.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                    PARAM_TEXT, '', VALUE_REQUIRED),
+        ]);
+    }
+
+    public static function function_returns() {
+        return new external_value(
+                PARAM_BOOL, '', VALUE_REQUIRED);
+    }
+
+    public static function function($params) {}
```
Comment (sliute) [inline tests/server_test.php]: This is meant to mirror the upsert() method in `auth/avadosalesforce/classes/external/upsert_v1.php`, right? But do we need to define it here inside this class? And if we do, can it use a more telling name, as 'function' is pretty generic?

File: tests/server_test.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            ->willReturn(['vfsStream' => 'life saver']);
+        $mockformat->expects($this->at(2))
+            ->method('send_headers');
+        $mockformat->expects($this->at(3))
+            ->method('send_response');
+
+        $server = new webservice_loadedrest_server_mock($mockformat, $inputfile->url());
+        $server->run();
+    }
+}
```
Comment (sliute) [inline tests/server_test.php:146]: And we need to add some assertions to this test, as there currently are none.

File: tests/xml_format_test.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        $this->assertRegExp('%\<exception%', $output);
+        $this->assertRegExp('%class="Exception"%', $output);
+        $this->assertRegExp('%code="1"%', $output);
+        $this->assertRegExp('%\<message\>message\</message\>%', $output);
+    }
+
+    public function data_send_response() {
+        return [
+            [
+                'result' => true,
```
Comment (sliute) [inline tests/xml_format_test.php:96]: Should we provide a key-value array as a value for 'result' here, like it happens below? As in, `'result' => ['key' => true]`.

File: tests/xml_format_test.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                    '%<key>3.14</key>%',
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider data_send_response
+     */
+    public function test_send_response($result, external_description $description, $expect) {
```
Comment (sliute) [inline tests/xml_format_test.php:132]: When running this test, I get this warning:

`1) Warning
The data provider specified for webservice_loadedrest_xml_format_testcase::test_send_response is invalid.
Class 'external_value' not found
`

Doesn't quite make sense, as `new external_value()` statements inside `data_send_response()` easily lead to `/lib/externallib.php` from inside PhpStorm.

**Review by sliute — COMMENTED** [review]
Looks fine, have just a few questions below.

File: tests/server_test.php (hunk 1 above)
Comment (LukeCarrier) [reply tests/server_test.php]: Yes in the sense that it's the implementation of a WS function, but no as they're separate plugins -- this is just a mock. If you have a better name I'll take it, but I'm keen to follow the naming convention in the [external functions API](https://docs.moodle.org/dev/External_functions_API).

File: tests/server_test.php (hunk 2 above)
Comment (LukeCarrier) [reply tests/server_test.php:146]: The test is intended to cover `server`'s interactions with `format`, which is achieved via the [PHPUnit `InvokedAtIndex` matcher](https://phpunit.de/manual/6.5/en/test-doubles.html#test-doubles.mock-objects.tables.matchers). I can't see any value in adding additional assertions?

File: tests/xml_format_test.php (hunk 3 above)
Comment (LukeCarrier) [reply tests/xml_format_test.php:96]: This is tested in the data block immediately below -- this one's to test `external_value`s in the root of the request.

File: tests/xml_format_test.php (hunk 4 above)
Comment (LukeCarrier) [reply tests/xml_format_test.php:132]: Can you paste the output of `$ vendor/bin/phpunit --group webservice_loadedrest` here? I'm interested in seeing the version information.

File: tests/server_test.php (hunk 2 above)
Comment (sliute) [reply tests/server_test.php:146]: OK, got it. It's just that when run, PHPUnit doesn't output any sort of status for those - it is like you had no tests whatsoever (no dots, no other notification etc.) But it does not fail.

File: tests/xml_format_test.php (hunk 4 above)
Comment (sliute) [reply tests/xml_format_test.php:132]: Here it is:

```
Moodle 3.4.2+ (Build: 20180405)
Php: 7.0.30.0.0.16.04.1, sqlsrv: 14.00.1000, OS: Linux 4.4.0-127-generic x86_64

PHPUnit 6.5.7 by Sebastian Bergmann and contributors.

......
```

File: tests/server_test.php (hunk 1 above)
Comment (sliute) [reply tests/server_test.php]: I'd go with `mock_upsert_parameters`, `mock_upsert_returns` and `mock_upsert`, but it's not a life-and-death issue.

**Review by sliute — APPROVED** [review]
That's it for now.

File: tests/xml_format_test.php (hunk 4 above)
Comment (LukeCarrier) [reply tests/xml_format_test.php:132]: Fixed via include of `externallib` -- it seems data providers execute prior to the PHPUnit bootstrap, which means the classloader isn't around to load the missing `external_*` classes.

File: tests/server_test.php (hunk 1 above)
Comment (LukeCarrier) [reply tests/server_test.php]: I'll meet you halfway with `mock_external_function`, since this is _not_ directly related to `auth_avadosalesforce`.

**Comment by LukeCarrier** [comment]
Fixed the above and added a `markTestSkipped()` for Moodle <= 3.3, which lack `vfsStream`.
